index.rst

```diff
@@ -13,17 +13,249 @@
+When the transfer job for that execution completes successfully, that journal file is removed.
+
+Changing the journal file format should be considered a data repository migration, and all migrations should require that the data repository have no active journal files unless they are able to migrate those files as well.
+
+Implementation of important butler operations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+``Butler.put``
+""""""""""""""
+
+As before, but a journal file will be written (sometime) before the ``Datastore.put`` begins and deleted after the Registry operation succeeds.
+
+``QuantumBackedButler`` will not write or delete journal files; it will rely entirely on a higher-level one for the full QuantumGraph.
+
+For a client/server Butler the journal creation and deletion could happen in the client or in the server, but the former continues to permit a regular ``FileDatastore`` to be used.
+
+``Butler.import`` and ``Butler.transfer_from``
+""""""""""""""""""""""""""""""""""""""""""""""
+
+As with ``put``, we would write a journal file before the Datastore operations begin and delete it after Registry writes succeed.
+
+``Butler.get``
+""""""""""""""
+
+No journaling is needed, as this is a read-only operation.
+
+``Butler.unstore`` and other removals
+"""""""""""""""""""""""""""""""""""""
+
+The journal file should be written before the ``Registry`` transaction is committed and deleted only after all Datastore deletions succeed.
+This is slightly problematic for client/server, because the journal file will need to be populated with information we get from the Registry database; this means the client cannot be responsible for creating the journal file unless we make fetching the datastore records and deleting them separate operations.
+That isn't too bad - it's just a slight increase in latency and a bit more http traffic.
+Another alternative would be to have the server take responsibility for creating the journal file, and then either returning responsibility for its deletion to the client or taking responsibility for both the deletion of the Datastore artifacts and the deletion of the journal file.
+Which of these is preferable probably depends on whether we want these operations to block until completion and whether we have reasons to perform other Datastore operations on the server (up to this point, having the client use a vanilla ``FileDatastore`` and perform all Datastore operations still seems viable).
+
+.. note::
+   TODO: think about how all of this interacts with DECam raws.
+
+.. note::
+   TODO: think about how all of this affects disassembled composites.
+   Or get rid of them!  I am pretty confident now that we'll never need them, and dropping that before we embark on other major changes to Datastore seems wise (sorry, Tim).
```

Comment:
We also talked about virtual composites. A dataset that could be assembled from components of multiple other datasets. This was how we thought we were going to create a "calibrated raw" by defining it to be a combination of the raw with the WCS, say, coming from the calexp, and the metadata coming from a recalculated header constructed from the EFD. This is also more efficient if the calexp WCS was stored as a separate component.

Reply:
I don't want to abandon composite disassembly without wider discussion with at least @ktlim and @gpdf. Now that we have read-only components we could change how we do disassembly such that we split it into data, variance, mask + ExposureInfo.

For the virtual composites with the updated headers calculated from the EFD, the idea originally was that we would do a batch job where we regenerate the header components from the EFD and store them as distinct datasets in butler. Then on `butler.get` it would understand that the `raw+` dataset type should be constructed by reading the raw then reading the header and then replacing the header component before returning it to the user. This seemed entirely like something that butler itself should be able to manage rather than asking people to retrieve the raw and then go looking for the relevant header dataset. The `raw++` dataset that has the raw pixel data plus all the final PSF and WCS information attached was also wanted at one point but maybe that could still be done without disassembly, but by downloading both the raw and the compressed PVI and then having a virtual composite define which components comes from the raw and which comes from the PVI (and you have to read both into memory if they are not disassembled).

Reply:
Ok, if the EFD information is getting there via a batch job that dumps the information into other butler datasets first, I'm much less concerned about it, as that batch job will have the kind of provenance I'm worried about.

And I don't think we need to be in a rush to drop composite assembly - I've been careful to leave room for datastore opaque tables that need more than just UUID (i.e. component) in their primary key, and I think that's the main potential concern here.

The DECam-raw case with multiple datasets in a single file is more worrisome, but mostly for deletes, I think, so if we declare that those can only be ingested via absolute (and hence "unowned") URIs, that may be fine, too.

Reply:
It sounds like the suggestion is that any generation of "virtual" data products including combinations of components (but not extraction of single components) will require a Pipeline, either with the user adding "converter" PipelineTasks to an existing Pipeline or behind the scenes of an image retrieval/regeneration service.  I'm a little worried about potential overheads of doing things this way, but exposing the conversion via a PipelineTask rather than hiding it in the Butler is not inherently unacceptable.

Reply:
> but mostly for deletes

Datastore currently has logic to handle the DECam case such that it only deletes the actual file if there are no more remaining records containing that file. This is easy when Datastore owns the records but sounds impossible if Registry is only telling Datastore what to delete without telling Datastore that other records exist for the file -- it seems that registry will have to have the logic of deciding for unstore whether to send any records to datastore at all, only sending records if all the records are relevant -- this then becomes problematic because then registry has to understand what a column in the records means. The whole point of Opaque is that registry doesn't know what the records mean in terms of real data files, or if indeed there are any files at all (which is another problem for signing -- the specific datastore is the only thing that knows that it has an actual file that can be converted to a public signed URL).

I imagine that in most cases "virtual" composites would be generated by pipelines somehow (or some batch system). The user cares about being able to retrieve something without knowing that it's virtual.

Reply:
To be a little more concrete about virtual composites, I was thinking along the lines of a new table that would have:

```
UUID | UUID1 | Component|
abc    | def      |                     |
abc    | xyz      | wcs             |
abc    | mno     | header        |
```
and you read this as being a virtual dataset known to registry as abc, with the entire content of def modified by replacing component wcs from xyz and component header from mno. Maybe we no longer need that sort of functionality. The StorageClass delegates we already have written and the component handling code in datastore make this fairly straightforward for reading.